README.md
# GitHub Actions MSBuild Logger

This is a custom `msbuild` and `dotnet` logger that, when run inside GitHub Actions, outputs warnings and errors to GitHub. This includes roslyn analyzers and code analysis.

## Integration

### dotnet

```
name: Build
on: [push]

jobs:
  dotnet:
    runs-on: windows-latest
    env:
      GitHubActionsLogger_Debug: True
    steps:
    - uses: actions/checkout@v1
    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1
    - uses: StanleyGoldman/setup-GitHubActionsMSBuildLogger@v1
    - name: Run dotnet
      run: |
        dotnet build -c Release .\TestConsoleApp1.sln /logger:GitHubActionsLogger,..\GitHubActionsMSBuildLogger.dll
```

### msbuild

```
name: Build
on: [push]

jobs:
  msbuild:
    runs-on: windows-latest
    env:
      GitHubActionsLogger_Debug: True
    steps:
    - uses: actions/checkout@v1
    - name: Setup MSBuild.exe
      uses: warrenbuckley/Setup-MSBuild@v1
    - uses: StanleyGoldman/setup-GitHubActionsMSBuildLogger@v1
    - name: Run msbuild
      run: |
        msbuild.exe -verbosity:minimal TestConsoleApp1.sln /logger:GitHubActionsLogger,..\GitHubActionsMSBuildLogger.dll
```


CHANGELOG.md
<a name="v1.0.21"></a>
## [v1.0.21](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.20...v1.0.21) - 2019-12-09

### Pull Requests
- Merge pull request [#18](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/18) from StanleyGoldman/add-license


<a name="v1.0.20"></a>
## [v1.0.20](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.19...v1.0.20) - 2019-12-09
### Refactor
- Adding the build version to the log output


### Pull Requests
- Merge pull request [#16](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/16) from StanleyGoldman/version-output


<a name="v1.0.19"></a>
## [v1.0.19](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.18...v1.0.19) - 2019-12-08
### Refactor
- Fixing the roslyn codeanalysis tests


### Pull Requests
- Merge pull request [#15](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/15) from StanleyGoldman/codeanalysis-deprecated


<a name="v1.0.18"></a>
## [v1.0.18](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.17...v1.0.18) - 2019-12-07
### Refactor
- Adding functionality for tests to run nuget


### Pull Requests
- Merge pull request [#14](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/14) from StanleyGoldman/fixing-roslynator


<a name="v1.0.17"></a>
## [v1.0.17](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.15...v1.0.17) - 2019-12-06

### Pull Requests
- Merge pull request [#9](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/9) from StanleyGoldman/dependabot/nuget/Microsoft.NET.Test.Sdk-16.4.0
- Merge pull request [#11](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/11) from StanleyGoldman/dependabot/nuget/xunit.runner.visualstudio-2.4.1


<a name="v1.0.15"></a>
## [v1.0.15](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.14...v1.0.15) - 2019-12-06

### Pull Requests
- Merge pull request [#13](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/13) from StanleyGoldman/branch-fix


<a name="v1.0.14"></a>
## [v1.0.14](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/compare/v1.0.0...v1.0.14) - 2019-12-06

### Pull Requests
- Merge pull request [#3](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/3) from StanleyGoldman/debugging
- Merge pull request [#2](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/2) from StanleyGoldman/nuget-package-fix
- Merge pull request [#1](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/1) from StanleyGoldman/nuget-org
- Merge pull request [#13](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/13) from StanleyGoldman/tests
- Merge pull request [#11](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/11) from StanleyGoldman/tests
- Merge pull request [#9](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/9) from StanleyGoldman/building-up
- Merge pull request [#7](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/7) from StanleyGoldman/dependabot/nuget/xunit-2.4.1
- Merge pull request [#4](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/4) from StanleyGoldman/dependabot/nuget/xunit.runner.visualstudio-2.4.1
- Merge pull request [#6](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/6) from StanleyGoldman/dependabot/nuget/coverlet.collector-1.1.0
- Merge pull request [#2](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/2) from StanleyGoldman/dependabot/nuget/Microsoft.NET.Test.Sdk-16.4.0
- Merge pull request [#3](https://github.com/StanleyGoldman/GitHubActionsMSBuildLogger/issues/3) from StanleyGoldman/actions


<a name="v1.0.0"></a>
## v1.0.0 - 2019-12-02



src/GitHubActionsMSBuildLogger.Tests/MSBuildTests.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using FluentAssertions;
using FluentAssertions.Execution;
using RunProcessAsTask;
using Xunit;
using Xunit.Abstractions;

namespace GitHubActionsMSBuildLogger.Tests
{
    public class MsBuildTests : BuildTestsBase
    {
        public MsBuildTests(ITestOutputHelper output)
        {
            _output = output;
            _msbuildExec = Environment.GetEnvironmentVariable("GitHubActionsMSBuildLogger_MsBuildPath");
            if (string.IsNullOrEmpty(_msbuildExec)) _msbuildExec = "msbuild";
        }

        private readonly ITestOutputHelper _output;
        private readonly string _msbuildExec;

        private async Task<ProcessResults> BuildAsync(string project, bool nugetRestore = false)
        {
            var loggerPath = GetLoggerPathOrThrow();
            var slnPath = GetSolutionPathOrThrow(project);

            if (nugetRestore)
            {
                await NugetRestoreAsync(slnPath)
                    .ConfigureAwait(false);
            }
            
            var version = await GetMsBuildVersionAsync()
                .ConfigureAwait(false);

            _output.WriteLine("msbuild version {0}", version);

            var processStartInfo =
                new ProcessStartInfo(_msbuildExec, $"/logger:GitHubActionsLogger,{loggerPath} {slnPath}")
                {
                    Environment =
                    {
                        {"GITHUB_ACTION", Guid.NewGuid().ToString()},
                        {"GITHUB_WORKSPACE", TargetPath}
                    }
                };

            return await ProcessEx.RunAsync(processStartInfo)
                .ConfigureAwait(false);
        }

        private async Task<string> GetMsBuildVersionAsync()
        {
            var processResults = await ProcessEx.RunAsync(new ProcessStartInfo(_msbuildExec, $"-version"))
                .ConfigureAwait(false);

            var regex = new Regex(@"^Microsoft \(R\) Build Engine version (.*?)\s.*$");
            var match = regex.Match(processResults.StandardOutput.First());
            
            return match.Groups[1].Value;
        }

        [Fact]
        public async Task TestCodeAnalysis()
        {
            using var processResults = await BuildAsync("codeanalysis")
                .ConfigureAwait(false);

            var (warnings, errors) = OutputResults(processResults, _output);

            using (new AssertionScope())
            {
                processResults.ExitCode.Should().Be(0);

                warnings.Should().BeEquivalentTo(
                    "::warning file=TestConsoleApp1,line=0,col=0:: Post-build Code Analysis (FxCopCmd.exe) has been deprecated in favor of FxCop analyzers, which run during build. Refer to https://aka.ms/fxcopanalyzers to migrate to FxCop analyzers.",
                    "::warning file=TestConsoleApp1/Program.cs,line=20,col=0::CA2213 Microsoft.Usage : 'Program.MyClass' contains field 'Program.MyClass._inner' that is of IDisposable type: 'Program.MyOTherClass'. Change the Dispose method on 'Program.MyClass' to call Dispose or Close on this field."
                );
                errors.Should().BeEmpty();
            }
        }

        [Fact]
        public async Task TestRoslynCodeAnalysis()
        {
            using var processResults = await BuildAsync("roslyn-codeanalysis", true)
                .ConfigureAwait(false);

            var (warnings, errors) = OutputResults(processResults, _output);

            using (new AssertionScope())
            {
                processResults.ExitCode.Should().Be(0);

                warnings.Should().BeEquivalentTo(
                    @"::warning file=TestConsoleApp1/Program.cs,line=15,col=0::CA1812 Program.MyClass is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic).",
                    @"::warning file=TestConsoleApp1/Program.cs,line=11,col=0::CA1801 Parameter args of method Main is never used. Remove the parameter or use it in the method body.",
                    @"::warning file=TestConsoleApp1/Program.cs,line=17,col=0::CA1823 Unused field '_inner'.",
                    @"::warning file=TestConsoleApp1/Program.cs,line=17,col=0::CA2213 'MyClass' contains field '_inner' that is of IDisposable type 'MyOTherClass', but it is never disposed. Change the Dispose method on 'MyClass' to call Close or Dispose on this field."
                    );
                errors.Should().BeEmpty();
            }
        }

        [Fact]
        public async Task TestRoslynator()
        {
            using var processResults = await BuildAsync("roslynator", true)
                .ConfigureAwait(false);

            var (warnings, errors) = OutputResults(processResults, _output);

            using (new AssertionScope())
            {
                processResults.ExitCode.Should().Be(0);

                warnings.Should().BeEquivalentTo(
                    @"::warning file=TestConsoleApp1/Program.cs,line=9,col=0::RCS1102 Make class static.", "::warning file=TestConsoleApp1/Program.cs,line=16,col=0::RCS1102 Make class static."
                );
                errors.Should().BeEmpty();
            }
        }

        [Fact]
        public async Task TestSimpleProjectError()
        {
            using var processResults = await BuildAsync("simple-project-error")
                .ConfigureAwait(false);

            var (warnings, errors) = OutputResults(processResults, _output);

            using (new AssertionScope())
            {
                processResults.ExitCode.Should().Be(1);

                warnings.Should().BeEmpty();
                errors.Should().BeEquivalentTo(
                    "::error file=TestConsoleApp1/Program.cs,line=13,col=0::CS1002 ; expected"
                );
            }
        }

        [Fact]
        public async Task TestSimpleProjectInfo()
        {
            using var processResults = await BuildAsync("simple-project-info")
                .ConfigureAwait(false);

            var (warnings, errors) = OutputResults(processResults, _output);

            using (new AssertionScope())
            {
                processResults.ExitCode.Should().Be(0);

                warnings.Should().BeEmpty();
                errors.Should().BeEmpty();
            }
        }

        [Fact]
        public async Task TestSimpleProjectWarning()
        {
            using var processResults = await BuildAsync("simple-project-warning")
                .ConfigureAwait(false);

            var (warnings, errors) = OutputResults(processResults, _output);

            using (new AssertionScope())
            {
                processResults.ExitCode.Should().Be(0);

                warnings.Should().BeEquivalentTo(
                    "::warning file=TestConsoleApp1/Program.cs,line=13,col=0::CS0219 The variable 'hello' is assigned but its value is never used"
                );
                errors.Should().BeEmpty();
            }
        }
    }
}